pkg/policy/policy.go

```diff
@@ -80,17 +79,40 @@ func (s *SearchContext) String() string {
 	}
 	for _, dport := range s.DPorts {
 		if dport.Name != "" {
-			dports = append(dports, fmt.Sprintf("%s/%s", dport.Name, dport.Protocol))
+			var str strings.Builder
+			str.Grow(len(dport.Name) + len(dport.Protocol) + 1)
+			str.WriteString(dport.Name)
+			str.WriteRune('/')
+			str.WriteString(dport.Protocol)
+			dports = append(dports, str.String())
 		} else {
-			dports = append(dports, fmt.Sprintf("%d/%s", dport.Port, dport.Protocol))
+			var str strings.Builder
+			dport_Str := strconv.FormatUint(uint64(dport.Port), 10)
+			str.Grow(len(dport_Str) + len(dport.Protocol) + 1)
+			str.WriteString(dport_Str)
+			str.WriteRune('/')
+			str.WriteString(dport.Protocol)
+			dports = append(dports, str.String())
 		}
 	}
-	ret := fmt.Sprintf("From: [%s]", strings.Join(from, ", "))
-	ret += fmt.Sprintf(" => To: [%s]", strings.Join(to, ", "))
+	from_str := strings.Join(from, ", ")
+	to_str := strings.Join(to, ", ")
+	dport_str := strings.Join(dports, ", ")
+	var ret_str strings.Builder
+	// the length of 'From: [] => To: [] Ports: []' is 28
+	ret_str.Grow(len(from_str) + len(to_str) + len(dport_str) + 28)
```

Comment:
Why do we need to add another 28? Where did 28 come from? I see you said 

```
the length of 'From: [] => To: [] Ports: []' is 28
```

but I'm still confused.

Reply:
i thought the final result string is something like:
`From: [strings.Join(from, ", ")] => To: [trings.Join(to, ", ")] Ports: [strings.Join(dports, ", ")]`

 So the length of `ret_str` is `len(strings.Join(from, ", ")) + len(strings.Join(to, ", ")) + len(strings.Join(dports, ", ")) + len('From: [] => To: [] Ports: []')` and the length of 'From: [] => To: [] Ports: []' is 28.

Reply:
Ah gotcha, makes sense.

Reply:
Ah gotcha, makes sense.